akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -736,66 +933,80 @@ private[akka] class Shard(
     if (hasBufferedMessages) {
       log.debug("Entity stopped after passivation [{}], but will be started again due to buffered messages", entityId)
-      // in case we're already writing to the remember store
       flightRecorder.entityPassivateRestart(entityId)
-      self ! StartEntityInternal(entityId)
+      // trigger start or batch in case we're already writing to the remember store
+      entities.rememberingStart(entityId, None)
+      if (!entities.pendingRememberedEntitiesExist()) rememberUpdate(Set(entityId))
     } else {
       log.debug("Entity stopped after passivation [{}]", entityId)
-      dropBufferFor(entityId)
     }
   }
 
-  /**
-   * @param entityIdsWaitingForWrite ids for remembered entities that have a write in progress, if non empty messages for that id
-   *                                will be buffered
-   */
-  private def deliverMessage(msg: Any, snd: ActorRef, entityIdsWaitingForWrite: OptionVal[Set[EntityId]]): Unit = {
-    val (id, payload) = extractEntityId(msg)
-    if (id == null || id == "") {
+  private def deliverMessage(msg: Any, snd: ActorRef): Unit = {
+    val (entityId, payload) = extractEntityId(msg)
+    if (entityId == null || entityId == "") {
       log.warning("Id must not be empty, dropping message [{}]", msg.getClass.getName)
       context.system.deadLetters ! msg
     } else {
       payload match {
         case start: ShardRegion.StartEntity =>
+          // Handling StartEntity both here and in the receives allows for sending it both as is and in an envelope
+          // to be extracted by the entity id extractor
+
           // we can only start a new entity if we are not currently waiting for another write
-          if (entityIdsWaitingForWrite.isEmpty) startEntities(Map(start.entityId -> Some(sender())))
-          // write in progress, see waitForAsyncWrite for unstash
-          else stash()
+          if (entities.pendingRememberedEntitiesExist()) {
+            if (VerboseDebug)
+              log.debug("StartEntity({}) from [{}], adding to batch", start.entityId, snd)
+            entities.rememberingStart(entityId, ackTo = Some(snd))
+          } else {
+            if (VerboseDebug)
+              log.debug("StartEntity({}) from [{}], starting", start.entityId, snd)
+            startEntity(start.entityId, Some(sender()))
+          }
         case _ =>
-          entities.entityState(id) match {
-            case OptionVal.Some(Remembering) =>
-              appendToMessageBuffer(id, msg, snd)
-            case OptionVal.Some(Passivating(_)) =>
-              appendToMessageBuffer(id, msg, snd)
-            case OptionVal.Some(Active(ref)) =>
-              if (VerboseDebug) log.debug("Delivering message of type [{}] to [{}]", payload.getClass, id)
-              touchLastMessageTimestamp(id)
+          entities.entityState(entityId) match {
+            case Active(ref) =>
+              // FIXME remember to remove payload toString
```

Comment:
now or later?

Reply:
I'll remove it now, was mostly for figuring out a specific bug earlier.